src/types/api.ts

```diff
@@ -0,0 +1,161 @@
+import { OktaAuth, OktaAuthOptions, Tokens } from '@okta/okta-auth-js';
+import { EventCallback, EventCallbackWithError, WidgetOptions } from './WidgetOptions';
+export interface HooksAPI {
+  before(eventName, hookFn): void;
+  after(eventName, hookFn): void;
+}
+export interface RouterEventsAPI {
+  on(event: 'ready', callback: EventCallback): void;
+  on(event: 'afterError', callback: EventCallbackWithError): void;
+  on(event: 'afterRender', callback: EventCallback): void;
+
+  off(event?: 'ready', callback?: EventCallback): void;
+  off(event?: 'afterError', callback?: EventCallbackWithError): void;
+  off(event?: 'afterRender', callback?: EventCallback): void;
+}
+
+export type RenderSuccessCallback = (res: RenderResult) => void;
+export type RenderErrorCallback = (err: RenderError) => void;
+export interface OktaSignInAPI extends HooksAPI, RouterEventsAPI {
+  authClient: OktaAuth;
+  show(): void;
+  hide(): void;
+  remove(): void;
+  showSignIn(options): Promise<RenderResult>;
+  showSignInToGetTokens(options: RenderOptions): Promise<Tokens>;
+  showSignInAndRedirect(options: RenderOptions): Promise<void>;
+  renderEl(
+    options: RenderOptions,
+    success?: RenderSuccessCallback,
+    error?: RenderErrorCallback
+  ): Promise<RenderResult>;
+
+  getUser(): void
+}
+export interface OktaSignInConstructor {
```

Comment:
Since OktaSignIn is already a `class`. why would still need an interface and constructor type?
- will be able to generate class type from the `class OktaSignIn` implementation?
- for developer, they property can do `const siw : OktaSignIn = new OktaSignIn({...})`?

Reply:
One downside of the class is that because it is both a type and value, there can be conflicts if you have two different versions of the object, if the newer version has added a property it won't be assignable to the old type. One way this could easily happen is because we allow the `authClient` to be passed in directly. We defend by declaring that authClient is using the OktaAuthInterface (the shape of AuthJS), not OktaAuth (the exact class we imported). The App may be using a different version of AuthJS than the widget was built against.  Or it is loading a different version widget than it had types for. In this case an app can do something like:

```
const OktaSignIn: OktaSignInConstructor = require('@okta/okta-signin-widget');
const signIn: OktaSignInAPI = new OktaSignIn(...options);
```

By using interfaces instead of classes, it is looser about the shape of the object that is returned from require() and new() - it only has to conform with the public API.